docs/src/density_interface.md

```diff
@@ -0,0 +1,21 @@
+# Support for DensityInterface
+
+`Distributions` supports [`DensityInterface`](https://github.com/JuliaMath/DensityInterface.jl) for distributions.
+
+For *single* variate values `x`, `DensityInterface.logdensityof(d::Distribution, x)` is equivalent to `logpdf(d, x)` and `DensityInterface.densityof(d::Distribution, x)` is equivalent to `pdf(d, x)`.
+
+You can use
```

Comment:
```suggestion
Additionally—though not supported by other packages using this API—you can use
```

Reply:
I don't think this should be added. Or at least I think it should be phrased differently. Currently it seems to suggest no other package supports and makes use of `IIDDensity` even though it is supposed to be used by DynamicPPL/Turing. It also only accepts a `Distribution`, so it doesn't seem to suggest that it can be used with arbitrary measures or densities. In the same way as eg a `RadonNikodymDerivative` probably would only support specific measures.

Reply:
You can export an `IIDDensity` from your package which uses Distributions. But you cannot call `IIDDensity` on an object outside the Distributions context. So if you are using `IIDDensity` on the importing side, you do not participate in `DensityInterface`, because you will restrict yourself to `Distribution`s and then you can just call `logpdf` instead and forget about `logdensityof`.

This is in contrast to the *semantics* of `IIDDensity` which (except the conflation of `value` and `[value]`) would work with any object that already supports `logdensityof` interface.

Reply:
> This is in contrast to the semantics of IIDDensity which (except the conflation of value and [value]) would work with any object that already supports logdensityof interface.

A default implementation wouldn't cover the implementation of `loglikelihood` since it does not assume that multiple samples are not necessarily arrays of individual samples. It's not relevant here if this is a good choice (I don't think so) but the only purpose of the (unexported!) `IIDDensity` is to be able to start the transition to DensityInterface if one relies on the behaviour of `loglikelihood`. It is still possible to support other densities by e.g. dispatching on `Distribution` and only using `IIDDensity` for them. Maybe in the long run `IIDDensity` won't be needed or useful, or maybe it will be upstreamed :shrug: But it allows to adopt DensityInterface without any breaking changes.

Reply:
> though not supported by other packages using this API

I also don't like this - it sound like `IIDDensity` is violating the contract of the `LogDensity`-API. But I think with the right interpretation/definition it doesn't: If we say it's the density over all possible outcomes of all possible draws from `d`, over all possible numbers of samples in each draw, then the behavior of `logdensityof` is just as defined, right?

Reply:
I would rather argue that the name of `Distributions.loglikelihood` isn't ideal ... but that's a different topic. :-)

Reply:
> I also don't like this - it sound like IIDDensity is violating the contract of the LogDensity-API. 

Ah, sorry @mschauer , I think I misunderstood - with "though not supported by other packages using this API" you were referring not to what `logdensityof` means for `IIDDensity` but you were referring to the relationship to `loglikelihood`?

Reply:
That's what I meant: the semantics if `IIDDensity` is as valid as the rest, it's just impossible to support `IIDDensity` without inheriting from `Distributions` because it is living here. So it's something a producer of densities using Distributions can use but not a consumer of density-like objects. It's perhaps ok like it is for now, I am just reviewing the PR because I care

Reply:
Ah right, sorry for the misunderstanding!

> just impossible to support IIDDensity without inheriting from Distributions because it is living here.

Yes, it is a bit of a weird thing, being in Distributions but being just a density. I guess we could, in principle turn IIDDensity into an actual `Distribution`: We can define `rand` for it (won't be very useful, but we can define it), and it's `logpdf` would be implemented by calling `loglikelihood` on the underlying Distribution. We wouldn't even need a special method for `logdensity` then. It wouldn't be possible to normalize the PDF though. Would that be acceptable?

Reply:
I think it would need to live outside `<:Distribution` because otherwise it breaks assumptions on how `Distribution`s behave, I am not even sure if defining a new variate form would do the trick. Compare that `Product` is only defined for univariate densities, not because products of multivariate densities don't make sense, just because `Product` was forced to take a place in the type hierarchy. Perhaps now it would even work with making product `ArrayLikeVariate{N} where N` without loosing too much, but that's a bit outside of our scope right now.

Reply:
Hm, if it was restricted to a number of draws/samples, we *could* normalize such a distribution. @devmotion , could you do something like this in Turing?

```julia
logdensityof(ProductOver(d, nsamples(x)), x)
```

There would need to be a function `nsamples(x)` that determines the number of "entries" in `x`, But that we should be able to code, right?

Reply:
I think a `ProductOver(d::Distribution, n::Integer)` could be very useful in general, even for a univariate `d`, because `Distributions.Product(FillArrays.Fill(d, n))` doesn't work (`rand` on it produces vectors filled with the same value in each entries).

Reply:
This will be fixed in my refactoring of `Product` (hopefully done within the next days), so it should be covered by the regular product distribution.

Reply:
That's great!

Reply:
> This will be fixed in my refactoring of Product

Awesome!